compat/optional.h

```diff
@@ -6,8 +6,15 @@
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
 
-#ifndef SQUID__SRC_BASE_OPTIONAL_H
-#define SQUID__SRC_BASE_OPTIONAL_H
+#ifndef SQUID__COMPAT_OPTIONAL_H
+#define SQUID__COMPAT_OPTIONAL_H
+
+#if __cplusplus > 201701L
+
```

Comment:
```C++
#if __cplusplus >= 201703L
```
Based on the linked godbolt tests and compiler documentation, this or similar conditions will only become true in Squid builds starting with [GCC](https://gcc.godbolt.org/z/7E8eneGee) v11. And it will not become true for [clang](https://gcc.godbolt.org/z/3YYbfG7jv) until we actually switch to C++17 (when this code becomes useless and should be removed).

I think that behavior does not match the [stated](https://github.com/squid-cache/squid/pull/1107/files#r933767132) goal of detecting "only outdated OS".

Reply:
Okay. Description change acceptable.

Reply:
> Okay. Description change acceptable.

I am not sure what description change you are talking about, but this change request is about the mismatch between the effects of the proposed _code_ and its stated (in the [comment](https://github.com/squid-cache/squid/pull/1107/files#r933767132) quoted below) goal of detecting "only outdated OS". This code does not detect outdated OSes. For example, this code may not enable std::optional use on the very latest OS release running clang (please see the change request for the details of what this code detects). If the intent is correct, then this code is wrong. If the intent is wrong, then I need to know what the correct intent is in order to review this code. If the tests this change request is based on are wrong, then the change request itself needs a fix. None of that is related to any "description change" though.

> https://github.com/squid-cache/squid/pull/1107/files#r933767132: What we gain with this model of portability is that bugs in our custom code is restricted to affecting only outdated OS.